### PR TITLE
Add `vexplain trace`

### PR DESCRIPTION
**Pull request by systay**
## Description
This PR introduces a new `vexplain trace` command that enhances query analysis by tracking data flow between operators during query execution. Instead of returning the normal results from a query, `vexplain trace` provides an execution plan with detailed runtime statistics, including:

* NoOfCalls: The number of times each operator was invoked during query execution.
* AvgNumberOfRows: The average number of rows returned per call for each operator.
* MedianNumberOfRows: The median number of rows returned per call for each operator.

## Example

```sql
-- Example query
VEXPLAIN TRACE SELECT
    o.o_orderpriority,
    EXTRACT(YEAR FROM o.o_orderdate) AS order_year,
    COUNT(DISTINCT o.o_orderkey) AS order_count,
    SUM(l.l_extendedprice * (1 - l.l_discount)) AS total_revenue
FROM
    orders o
JOIN
    lineitem l ON o.o_orderkey > l.l_orderkey
WHERE
    o.o_orderdate BETWEEN '1995-01-01' AND '1996-12-31'
GROUP BY
    o.o_orderpriority,
    EXTRACT(YEAR FROM o.o_orderdate)
ORDER BY
    o.o_orderpriority,
    order_year
```

Here is the output of that query:

```json
{
    "OperatorType": "Aggregate",
    "Variant": "Ordered",
    "NoOfCalls": 1,
    "AvgNumberOfRows": 10,
    "MedianNumberOfRows": 10,
    // ... other fields ...
    "Inputs": [
        {
            "OperatorType": "Sort",
            "Variant": "Memory",
            "NoOfCalls": 1,
            "AvgNumberOfRows": 1995,
            "MedianNumberOfRows": 1995,
            // ... other fields ...
            "Inputs": [
                {
                    "OperatorType": "Route",
                    "Variant": "Scatter",
                    "NoOfCalls": 2,
                    "AvgNumberOfRows": 1000,
                    "MedianNumberOfRows": 1000,
                    // ... other fields ...
                },
                {
                    "OperatorType": "Route",
                    "Variant": "Scatter",
                    "NoOfCalls": 2000,
                    "AvgNumberOfRows": 1.995,
                    "MedianNumberOfRows": 2,
                    // ... other fields ...
                }
            ]
        }
    ]
}
```

## Motivation
Understanding the internal workings of query execution is crucial for performance tuning and optimization. While traditional `VEXPLAIN` commands show the planned execution path, they lack runtime statistics. The `vexplain trace` command fills this gap by providing real-time data flow information, allowing for deeper insights into query performance and helping to identify bottlenecks more effectively.

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by GuptaManan100 — APPROVED** [review]
LGTM!

File: go/vt/vtgate/engine/fake_vcursor_test.go [hunk 1]
```diff
@@ -874,6 +878,10 @@ func (f *loggingVCursor) UnresolvedTransactions(_ context.Context, _ string) ([]
 	return f.transactionStatusOutput, nil
 }
 
+func (f *loggingVCursor) StartPrimitiveTrace() func() map[int]RowsReceived {
+	panic("implement me")
+}
+
 // SQLParser implements VCursor
 func (t *loggingVCursor) SQLParser() *sqlparser.Parser {
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/fake_vcursor_test.go:886]: nit: This is not needed when `noopVCursor` has same implementation.

File: go/vt/vtgate/engine/primitive.go [hunk 2]
```diff
@@ -260,6 +267,12 @@ type (
 
 	// txNeeded is a default implementation for Primitives that need transaction handling
 	txNeeded struct{}
+
+	PrimitiveID int
+
+	identifiablePrimitive struct {
+		id PrimitiveID
+	}
 )
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/primitive.go:276]: I think we can get away without having this to be added to all the primitives and can use primitive address pointer to tract them.

File: go/vt/vtgate/engine/trace.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+var _ Primitive = (*Trace)(nil)
+
+type Trace struct {
+	identifiablePrimitive
+	Inner Primitive
+}
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/trace.go:34]: nit: should this be made part of vexplain primitive? execute the code based on the variant

File: go/vt/vtgate/vcursor_impl.go [hunk 4]
```diff
@@ -645,6 +690,7 @@ func (vc *vcursorImpl) ExecuteStandalone(ctx context.Context, primitive engine.P
 	// The autocommit flag is always set to false because we currently don't
 	// execute DMLs through ExecuteStandalone.
 	qr, errs := vc.executor.ExecuteMultiShard(ctx, primitive, rss, bqs, NewAutocommitSession(vc.safeSession.Session), false /* autocommit */, vc.ignoreMaxMemoryRows, vc.resultsObserver)
+	vc.logOpTraffic(primitive, qr)
 	return qr, vterrors.Aggregate(errs)
 }
```
Comment (harshit-gangal) [inline go/vt/vtgate/vcursor_impl.go:695]: here logging will happen twice as this calls back into ExecuteMultiShard which will log again.

Comment (harshit-gangal) [reply go/vt/vtgate/vcursor_impl.go:695]: Ignore the above statement

**Review by harshit-gangal — COMMENTED** [review]
There is an extra logging incase of retry. As this is recorded for testing in a controlled way, can be ignored.

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**